### PR TITLE
GMAC implementation.

### PR DESCRIPTION
**Pull request by paulidale**
Implementation of the GMAC algorithm using the new EVP_MAC APIs.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/gmac/build.info [hunk 1]
```diff
@@ -0,0 +1,2 @@
+LIBS=../../libcrypto
+SOURCE[../../libcrypto]=gmac.c
```
Comment (levitte) [inline crypto/gmac/build.info]: Oh, a method only implementation?  Now, *that's* rad!  I like!  :smiley:

File: crypto/gmac/gmac.c [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    else if (strcmp(type, "hexkey") == 0)
+        return EVP_hex2ctrl(gmac_ctrl_str_cb, gctx, EVP_MAC_CTRL_SET_KEY,
+                            value);
+    else if (strcmp(type, "iv") == 0)
+        return EVP_str2ctrl(gmac_ctrl_str_cb, gctx, EVP_MAC_CTRL_SET_IV,
+                            value);
+    else if (strcmp(type, "hexiv") == 0)
+        return EVP_hex2ctrl(gmac_ctrl_str_cb, gctx, EVP_MAC_CTRL_SET_IV,
+                            value);
+    return -2;
```
Comment (levitte) [inline crypto/gmac/gmac.c]: It should be possible to set the cipher as well, no?  CMAC has that possibility...

File: doc/man7/EVP_MAC_GMAC.pod [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+=head2 Supported controls
+
+The supported controls are:
+
+=over 4
+
+=item B<EVP_MAC_CTRL_SET_KEY>
+
+EVP_MAC_ctrl_str() takes to type string for this control:
```
Comment (levitte) [inline doc/man7/EVP_MAC_GMAC.pod]: s/to/two/

File: doc/man7/EVP_MAC_GMAC.pod [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+=item "hexkey"
+
+The value string is expected to be a hexadecimal number, which will be
+decoded before passing on as control value.
+
+=back
+
+=item B<EVP_MAC_CTRL_SET_IV>
+
+EVP_MAC_ctrl_str() takes to type string for this control:
```
Comment (levitte) [inline doc/man7/EVP_MAC_GMAC.pod]: s/to/two/

File: doc/man7/EVP_MAC_GMAC.pod [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+=back
+
+=item B<EVP_MAC_CTRL_SET_ENGINE>
+
+=item B<EVP_MAC_CTRL_SET_CIPHER>
+
+These work as described in L<EVP_MAC(3)/CONTROLS> with the restriction that the
+cipher must be an AEAD one.
+
+EVP_MAC_ctrl_str() type string for B<EVP_MAC_CTRL_SET_CIPHER>: "cipher"
```
Comment (levitte) [inline doc/man7/EVP_MAC_GMAC.pod]: Ah, so you do have that intention.  Good!

File: crypto/gmac/build.info (hunk 1 above)
Comment (paulidale) [reply crypto/gmac/build.info]: No legacy support this way :)

File: crypto/gmac/gmac.c (hunk 2 above)
Comment (paulidale) [reply crypto/gmac/gmac.c]: Oops, I missed that bit.  I do allow the cipher via the vararg call.  I'll add.

**Comment by paulidale** [comment]
I'd prefer to not have a PKEY version of this if possible.  We need something to entice people over to the new API :)

**Comment by paulidale** [comment]
It isn't clear to me what has failed in the Travis runs.  It is a test failure but I can't see the test that failed.  It doesn't look to be the evpmac tests which are the new ones.

**Comment by levitte** [comment]
> I'd prefer to not have a PKEY version of this if possible. We need something to entice people over to the new API :)

Fine by me

File: test/evp_test.c [hunk 6]
```diff
@@ -936,6 +939,8 @@ static int mac_test_parse(EVP_TEST *t,
 
     if (strcmp(keyword, "Key") == 0)
         return parse_bin(value, &mdata->key, &mdata->key_len);
+    if (strcmp(keyword, "IV") == 0)
+        return parse_bin(value, &mdata->iv, &mdata->iv_len);
```
[the remark this answers is not in the document]
Comment (paulidale) [reply test/evp_test.c]: Thanks for that.  Fixed now.

File: crypto/gmac/gmac.c [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    const unsigned char *p;
+    size_t len;
+    EVP_CIPHER_CTX *ctx = gctx->ctx;
+    const EVP_CIPHER *cipher;
+    ENGINE *engine;
+
+    switch (cmd) {
+    case EVP_MAC_CTRL_SET_CIPHER:
+        cipher = va_arg(args, const EVP_CIPHER *);
+        return (EVP_CIPHER_flags(cipher) & EVP_CIPH_FLAG_AEAD_CIPHER)
```
[the remark this answers is not in the document]
Comment (paulidale) [reply crypto/gmac/gmac.c]: :)

Without any alternative return is this really required?
I'm happy to add it and should probably add `!= 0` to the end regardless.

[the remark this answers is not in the document]
Comment (paulidale) [reply crypto/gmac/gmac.c]: I am also leaving open the possibility of non-GCM AEAD ciphers being allowed.  I'm not too concerned but some people might consider this detrimental.  Is there anyway to check for a GCM cipher without parsing the cipher name?

File: crypto/gmac/gmac.c [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+                && EVP_EncryptInit_ex(ctx, cipher, NULL, NULL, NULL);
+
+    case EVP_MAC_CTRL_SET_KEY:
+        p = va_arg(args, const unsigned char *);
+        len = va_arg(args, size_t);
+        return len == (size_t)EVP_CIPHER_CTX_key_length(ctx)
```
[the remark this answers is not in the document]
Comment (paulidale) [reply crypto/gmac/gmac.c]: Yeah, this deserves a better error return.  Again there is no alternative to confuse users.
I don't remember what CMAC did in this case.

**Comment by paulidale** [comment]
I'm also not supporting truncation of the MAC output.  This is allowed but strongly discouraged by NIST (rightfully so).

**Comment by paulidale** [comment]
PS: Shane is looking into KMAC which should further stress the EVP MAC API -- it supports both fixed and variable length output.

File: crypto/gmac/gmac.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (paulidale) [reply crypto/gmac/gmac.c]: BTW: both elegance and gross inelegance are trademarks of mine 👍 
Unfortunately, I tend more towards the latter.  After all the code is obvious....

[the remark this answers is not in the document]
Comment (levitte) [reply crypto/gmac/gmac.c]: On your earlier blathering (I saw it!), we must also keep in mind that the user doesn't know if there's just one thing that can go wrong or not.

> Is there anyway to check for a GCM cipher without parsing the cipher name?

I have no idea. Unless there's some other indication of mode that I don't know about, errrr no. Or well, there's hard-coding the nids...

File: doc/man7/EVP_MAC_GMAC.pod [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+=head2 Supported controls
+
+The supported controls are:
+
+=over 4
+
+=item B<EVP_MAC_CTRL_SET_KEY>
+
+EVP_MAC_ctrl_str() takes two type string for this control:
```
Comment (levitte) [inline doc/man7/EVP_MAC_GMAC.pod]: In correcting my own, I realised that "string" should be "strings"

File: doc/man7/EVP_MAC_GMAC.pod [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+=item "hexkey"
+
+The value string is expected to be a hexadecimal number, which will be
+decoded before passing on as control value.
+
+=back
+
+=item B<EVP_MAC_CTRL_SET_IV>
+
+EVP_MAC_ctrl_str() takes two type string for this control:
```
Comment (levitte) [inline doc/man7/EVP_MAC_GMAC.pod]: Here as well

File: crypto/gmac/gmac.c [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+                            value);
+    if (strcmp(type, "iv") == 0)
+        return EVP_str2ctrl(gmac_ctrl_str_cb, gctx, EVP_MAC_CTRL_SET_IV,
+                            value);
+    if (strcmp(type, "hexiv") == 0)
+        return EVP_hex2ctrl(gmac_ctrl_str_cb, gctx, EVP_MAC_CTRL_SET_IV,
+                            value);
+    if (strcmp(type, "hexiv") == 0)
+        return EVP_hex2ctrl(gmac_ctrl_str_cb, gctx, EVP_MAC_CTRL_SET_IV,
+                            value);
```
Comment (kroeckx) [inline crypto/gmac/gmac.c]: hexiv seems to be duplicate

**Comment by paulidale** [comment]
I've noticed that the CIs build fast Sunday afternoon/evening in the US :)

**Comment by paulidale** [comment]
Merged, thanks everyone.

**Pull request closed without merging**